### PR TITLE
Feature/dapp service worker end to end tests

### LINKED ISSUE #2837: dApp - extend end-to-end tests to partially test service worker functionality
This [this comment](https://github.com/raiden-network/light-client/issues/2818#issuecomment-868542268) of #2818 at `End-to-end tests`.

### PR DESCRIPTION
**Pull request by weilbith**
**DON'T BE AFRAID OF THE DIFF SIZE.**
The vast majority of it comes from the `yarn.lock` file.

Fixes #2837

**Short description**
This only tests the offline functionality of the dApp as provided by the service worker. See the linked issue for more details. There will be a follow up issue/PR for further tests.

In fact this PR does only serve the dApp for the Cypress tests manually (before it was a plugin) and stop doing so after a timeout. From then on it must run offline. If the tests don't pass anymore, the offline functionality is not working. This can be tested by trying to make it work without the service worker. Therefore this test is actually working and verifies this functionality.

I'm not too happy with the solution of a bash script to orchestrate this whole process. But as the according commit message describes, this was the final best option. I tried many other alternative approaches, but none actually worked. So I think this is good enough. It does what it should reliably and very simple.

**Definition of Done**

- [x] Steps to manually test the change have been documented
- [x] Acceptance criteria are met
- [ ] Change has been manually tested by the reviewer (dApp) 

**Steps to manually test the change (dApp)**

1. Nothing really... just let the CI pipeline do its job.

### CODE REVIEW

**Comment by weilbith** [comment]
No CHANGELOG entry necessary for this PR>

**Comment by weilbith** [comment]
Somehow I messed up the dependencies. I'll fix that part...

**Comment by weilbith** [comment]
Alright, the last commit which is a rebase on `master` has 20 consecutive successful runs.
The CI workflow has run 107 over the weekend to exclude flakiness. Only two times the dApp end-to-end test failed for a timing issue, which has been addresses with a recent commit to prevent/reduce its likelihood.
All other failing runs where either the SDK end-to-end tests (was addressed and merged to `master`, since rebase on `master` it did not happen again) or twice the dApp unit tests failed due to a timing issue (addressed as well in recent commits, since that it did not happen again).

I think its finally time to review and merge this PR.

File: raiden-dapp/run-cypress.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+function ensure_cypress_is_available() {
+  npx cypress install
+}
+
+# This includes an time saving improvement to re-use the latest build if it
+# available. This is especially helpful for the CI workflow. Mind that this does
+# not evaluate for which mode the build was made for.
+function build_dapp_if_necessary() {
+  if [[ ! -d "$BUILD_FOLDER" ]]; then
```
Comment (weilbith) [inline raiden-dapp/run-cypress.sh:23]: Only do this for CI.

File: raiden-dapp/run-cypress.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+# The serve process will run in the background and gets killed when the script
+# ends (in fact when Cypress finished running the tests).
+# After a sensible timeout it will delete all asset files that get served. This
+# makes them unavailable for any client. This is used to verify that the dApp
+# works fully offline once it cached all assets. Though the serve process must
+# remain active as Cypress will fail else. This is related to to a code coverage
+# collection feature that can't be disabled.
+function serve_dapp_and_delete_assets_after_a_while() {
+  local temporally_serve_folder="$(mktemp -d)"
```
Comment (weilbith) [inline raiden-dapp/run-cypress.sh:36]:
```suggestion
   local temporary_serve_folder="$(mktemp -d)"
```

File: raiden-dapp/src/model/types.ts [hunk 3]
```diff
@@ -94,6 +94,7 @@ export const PlaceHolderNetwork: providers.Network = {
 };
 
 export enum ErrorCode {
+  DEPLOYMENT_INFO_PARSING_FAILED = 'deployment-info-parsing-failed',
```
Comment (weilbith) [inline raiden-dapp/src/model/types.ts:97]: Where is the catch block? (missed to commit it?)

File: raiden-dapp/tests/e2e/support/commands.ts [hunk 4]
```diff
@@ -16,9 +16,9 @@ Cypress.Commands.add(
   },
   (subject, selector) => {
     if (subject) {
-      cy.get(subject).get(selector, { timeout: 1800000 });
+      cy.get(subject).get(selector, { timeout: 180000 });
     } else {
-      cy.get(selector, { timeout: 1800000 });
+      cy.get(selector, { timeout: 180000 });
```
Comment (weilbith) [inline raiden-dapp/tests/e2e/support/commands.ts:21]: Improve readability with `3 * 60 * 1000`

File: raiden-dapp/tests/e2e/utils/navigation.ts [hunk 5]
```diff
@@ -166,5 +166,18 @@ export function navigateToDownloadLogs() {
   // cypress selectors: raiden-dapp/src/views/account/AccountRoot.vue
   cy.get('[data-cy=account_root]').should('exist');
   // cypress selectors: raiden-dapp/src/components/account/AccountContent.vue
-  cy.get('[data-cy=account_content_menu_list_items]').eq(4).click();
+  cy.get('[data-cy=account_content_menu_list_items]').eq(2).click();
+}
+
+/*
+ *
+ */
```
Comment (weilbith) [inline raiden-dapp/tests/e2e/utils/navigation.ts:189]: Remove

File: raiden-dapp/tests/e2e/specs/scenario.spec.ts [hunk 6]
```diff
@@ -96,7 +97,8 @@ describe('dApp e2e tests', () => {
     navigateToUDC();
     enterUDCWithdrawalAmount(uiTimeout);
     withdrawUDCTokens();
-    navigateBackToAccountMenu();
+    reloadWholeApplication();
```
Comment (weilbith) [inline raiden-dapp/tests/e2e/specs/scenario.spec.ts:100]: Maybe we could add a logging here to see when this happens timewise in relation to the server.

File: raiden-dapp/run-cypress.sh [hunk 7]
```diff
@@ -39,6 +39,7 @@ function serve_dapp_and_delete_assets_after_a_while() {
   cp $DEPLOYMENT_SERVICES_INFO "$temporally_serve_folder/"
   npx serve "$temporally_serve_folder" -p 5000 &
   SERVE_PID=$!
+  sleep 5s # The npx command might have to first download the serve application
```
Comment (weilbith) [inline raiden-dapp/run-cypress.sh:42]: Maybe we could first just install `serve` and then run it. More explicit.

**Review by weilbith — COMMENTED** [review]
This is the result of an interactive review with @palango . (I'm writing this review in representation of him, as we did the review on my PC).

File: raiden-dapp/tests/e2e/specs/scenario.spec.ts (hunk 6 above)
Comment (weilbith) [reply raiden-dapp/tests/e2e/specs/scenario.spec.ts:100]: Atm I don't know how to get a relative time value here....

File: raiden-dapp/tests/e2e/utils/navigation.ts (hunk 5 above)
Comment (weilbith) [reply raiden-dapp/tests/e2e/utils/navigation.ts]: ✔️

File: raiden-dapp/run-cypress.sh (hunk 1 above)
Comment (weilbith) [reply raiden-dapp/run-cypress.sh]: ✔️

File: raiden-dapp/run-cypress.sh (hunk 2 above)
Comment (weilbith) [reply raiden-dapp/run-cypress.sh]: ✔️

File: raiden-dapp/tests/e2e/support/commands.ts (hunk 4 above)
Comment (weilbith) [reply raiden-dapp/tests/e2e/support/commands.ts]: ✔️

File: raiden-dapp/run-cypress.sh (hunk 7 above)
Comment (weilbith) [reply raiden-dapp/run-cypress.sh]: ✔️

File: raiden-dapp/src/model/types.ts (hunk 3 above)
Comment (weilbith) [reply raiden-dapp/src/model/types.ts:97]: ✔️

**Comment by weilbith** [comment]
I addressed most of the feedback, except the time logging where I don't know yet how to do it. In the time I hope that CI can verify that the changes where not harmful.

File: raiden-dapp/tests/e2e/specs/scenario.spec.ts (hunk 6 above)
Comment (palango) [reply raiden-dapp/tests/e2e/specs/scenario.spec.ts:100]: Not that important.

**Review by palango — APPROVED** [review]
Great work, and the walk through review was really nice and helpful!

**Comment by weilbith** [comment]
Alright. Fixes are working. I squash all remaining fixup commits, trigger the pipe a couple of times in parallel and then merge.

**Comment by weilbith** [comment]
Alright. 12 confirming CI runs on `HEAD`.
As I said, the CHANGELOG must not be extended. And CodeCov is stupid...

Btw: the average time for the CI pipeline dropped by ~2min.

**Pull request merged**